aiida/engine/processes/calcjobs/monitors.py

```diff
@@ -0,0 +1,196 @@
+# -*- coding: utf-8 -*-
+"""Utilities to define monitor functions for ``CalcJobs``."""
+from __future__ import annotations
+
+import collections
+import dataclasses
+from datetime import datetime, timedelta
+import enum
+import inspect
+import typing as t
+
+from aiida.common.lang import type_check
+from aiida.common.log import AIIDA_LOGGER
+from aiida.orm import CalcJobNode, Dict
+from aiida.plugins import BaseFactory
+from aiida.transports import Transport
+
+LOGGER = AIIDA_LOGGER.getChild(__name__)
+
+
+class CalcJobMonitorAction(enum.Enum):
+    """The action a engine should undertake as a result of a monitor."""
+
+    KILL = 'kill'  # The job should be killed and no other monitors should be called.
+    DISABLE_ALL = 'disable-all'  # All monitors should be disabled for the continued duration of the current job
+    DISABLE_SELF = 'disable-self'  # Disable the monitor that returns a result with this action.
+
+
+@dataclasses.dataclass
+class CalcJobMonitorResult:
+    """Data class representing the result of a monitor."""
+
+    key: str | None = None
+    """Key of the monitor in the ``monitors`` input namespace. Will be set automatically by the engine."""
+
+    message: str | None = None
+    """Human readable message: could be a warning or an error message."""
+
+    action: CalcJobMonitorAction = CalcJobMonitorAction.KILL
+    """The action the engine should take."""
+
+    retrieve: bool = True
+    """If set to ``False``, the engine will skip retrieving the output files."""
+
+    parse: bool = True
+    """If set to ``False``, the engine will skip the parsing of the retrieved files, if one was specified in inputs."""
+
+    override_exit_code: bool = True
+    """If set to ``False``, the engine will keep the exit code returned by the parser."""
+
+    def __post_init__(self):
+        """Validate the attributes."""
+        self.validate()
+
+    def validate(self):
+        """Validate the instance.
+
+        :raises TypeError: If any of the attributes are of the incorrect type.
+        :raises ValueError: If ``parse == True`` and ``retrieve == False``.
+        """
+        type_check(self.key, str, allow_none=True)
+        type_check(self.message, str, allow_none=True)
+        type_check(self.action, CalcJobMonitorAction)
+        type_check(self.retrieve, bool)
+        type_check(self.parse, bool)
+        type_check(self.override_exit_code, bool)
```

Comment:
I did, but I don't want to add a dependency just for this.

Reply:
Fair enough. I think it might be worth considering to do more type checking in general in other sections of the code or move the ones that already exist to Pydantic, so it wouldn't be just for this, but I guess that can be a separate issue.